### ISSUE TITLE
Guide on custom plugins?

### ISSUE THREAD

**Issue by delaneyj** [hayes/giraphql#13]
I have a situation where I need to inject a minimum set of properties no matter what was queried. So 
```gql
{
  foo{
    bar
  }
}
```
would become 
```gql
{
  foo{
    id
    metadata
    bar
  }
}
```

I'm guessing a middleware style plugin would be great.  Is there a guide of what would be necessary to create own plugins that work with giraphql?

**Comment by hayes** [comment]
Hey, thanks for the question.

Unfortunately, I think my answer won't be exactly what you are looking for.

I have not gotten around to writing docs on plugins just yet, and part of what is missing is what plugins in GiraphQL can and cannot do.  What you are asking about is unfortunately something that is not possible in GiraphQL as it exists today. 

To explain why, its important to understand how graphql services generally work, and what role GiraphQL plays in that.

Most node graphql server implementations work by wrapping the graphql js implementation, and using it's query executor, or by using a custom executor like graphql-jit.  Either way, there are generally 2 somewhat separate piece at play.  The schema, which defines the shape of the graph, and all the methods to resolve fields/values in the graph, and the executor that takes a query and figures out what parts of the schema to execute to generate a response.

GiraphQL currently only covers the schema portion I mentioned above, this means it describes the graph, and how values are resolved, but the part that actually decides what resolves are called, and stitches everything together into a single response object is completely outside of the control of giraphql.  I have been thinking about adding a query executor to giraphql so that it is a more complete solution, but for now, those are the limitations that exist. 

If you want to inject things into a query that are not explicitly queried, that somewhat violates the core assumptions about how graphql works, and can cause issues such as collisions if a query uses aliases of the same name as the fields you are injecting.  With that warning out of the way, there are a few ways you can do this, the specifics will depend on what server/query executor you are using. The easiest way I can think to accomplish what you want would be to parse the incoming query into an a graphql document (AST) and then inject selections for the fields you want to query, and then either pass that modified query document.


I know this is a bit vague, but its hard to give a more specific answer without understanding a lot about your use case.  One thing that may be worth considering is if it would make more sense to add those fields into the query on the client.  This is a common pattern used by relay and apollo clients to add in fields that they require for things like caching and refeches.

**Comment by hayes** [comment]
Documentation for the plugin API will be part of the 2.0 release.  Work for 2.0 is being tracked in #32

**Comment by hayes** [comment]
New guide for writing plugins has been released as part of 2.0:  https://giraphql.com/guide/writing-plugins
